dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,298 @@
+        )
+        if not self.verbose:
+            output_df.drop(
+                labels=self._output_column_names.error_raw, axis=1, inplace=True)
+        if self.error_handling == ErrorHandling.FAIL:
+            error_columns = [
+                self._output_column_names.error_message,
+                self._output_column_names.error_type,
+                self._output_column_names.error_raw,
+            ]
+            output_df.drop(labels=error_columns, axis=1,
+                           inplace=True, errors="ignore")
+        return output_df
+
+    def run(self, df: pd.DataFrame, **function_kwargs,) -> pd.DataFrame:
+        """Applies a function to a pandas.DataFrame with parallelization, error logging and progress tracking.
+
+        The DataFrame is iterated on and fed to the function as dictionaries, row-by-row or by batches of rows.
+        This process is accelerated by the use of concurrent threads and is tracked with a progress bar.
+        Errors are catched if they match the `self.exceptions_to_catch` attribute and automatically logged.
+        Once the whole DataFrame has been iterated on, results and errors are added as additional columns.
+
+        Args:
+            df: Input dataframe on which the function will be applied
+            **function_kwargs: Arbitrary keyword arguments passed to the `function`
+
+        Returns:
+            Input dataframe with additional columns:
+            - response from the `function`
+            - error message if any
+            - error type if any
+
+        """
+        # First, we create a generator expression to yield each row of the input dataframe.
+        # Each row will be represented as a dictionary like {"column_name_1": "foo", "column_name_2": 42}
+        df_row_generator = (index_series_pair[1].to_dict()
+                            for index_series_pair in df.iterrows())
+        # We then chunk the generator into lists of rows of len batch_size
+        df_row_batch_generator = chunked(df_row_generator, self.batch_size)
+        df_num_rows = len(df.index)
+        len_generator = math.ceil(df_num_rows / self.batch_size)
+        logging.info(
+            f"Applying function {self.function.__name__} in parallel to {df_num_rows} row(s)"
+            + f" using batch size of {self.batch_size}..."
+        )
+        start = perf_counter()
+        self._output_column_names = self._get_unique_output_column_names(
+            existing_names=df.columns)
+        pool_kwargs = function_kwargs.copy()
+        for kwarg in ["function", "row", "batch"]:  # Reserved pool keyword arguments
+            pool_kwargs.pop(kwarg, None)
+        results = []
+        with ThreadPoolExecutor(max_workers=self.parallel_workers) as pool:
+            futures = [
+                pool.submit(self._apply_function_and_parse_response,
+                            batch=batch, **pool_kwargs)
+                for batch in df_row_batch_generator
+            ]
```

Comment:
What is the benefit of switching from a list comprehension to a for loop? The number of lines is the same.

Reply:
I find it clearer as it reads from left to right and not the opposite and makes use of indent

Reply:
Appreciate the point about clarity. As you mentioned it, what is the cost in terms of speed? Is it worth it?

Reply:
I think that the time increase is insignificant compared to the time to call an external api

Reply:
That is an interesting point. To be sure, I ran a small experiment: https://colab.research.google.com/drive/1rrHQcUMBrfHFvfqwcset47lZFj-2VL_q?usp=sharing 
And found this more full-fledged notebook: https://anaconda.org/afalkovskiy/comp_time_compare/notebook
You are right, for long running functions, the time difference is -1% between a for loop and a list comprehension.

So the question boils down to readability. I don't have a strong opinion, only a long-standing habit to use list comprehensions, because it is close to the set theory notation I am used to in statistics. I appreciate that list comprehensions are not possible in languages such as javascript and less common than for loops. So they can be less readable.

I have switched to a for loop here: 8d1aaa0

Reply:
Thanks for the experiment and for switching to a for loop ! 🙏